param.go

```diff
@@ -98,9 +98,36 @@ type paramVisitor interface {
 // recursed into.
 type paramVisitorFunc func(param) (recurse bool)
 
-func (f paramVisitorFunc) Visit(p param) paramVisitor {
-	if f(p) {
-		return f
+// paramVisitOnce encapsulates a paramVisitorFunc and a map to track visited nodes.
+type paramVisitOnce struct {
+	f       paramVisitorFunc
+	visited map[string]struct{}
```

Comment:
It would be safer to use the `key` representation here instead of `String()`.
I don't like the idea of making the human-readable representation part of core
business logic.

```
map[key]struct{}
```

And,

```
case paramSingle:
  k = key{name: p.Name, t: p.Type}
case paramGroup:
  k = key{group: p.Group, t: p.Type.Elem()}  // []Foo -> Foo
```

Reply:
This is a good idea. I didn't love `.String()`, but I was unable to use `p` directly as I don't believe it's hashable. Let me construct a `key`. Actually, give how many times `key`s are constructed now, does it make sense to make that a method on the various `param` types?

Reply:
It does, but I would skip it. I had been working on removing the `key` type
completely when this code was being refactored. The semantics with `key` get
weird because only leaf nodes have `key`s.